File: src/GeoTools/GeoTile.cpp
//
// Created by pshivaraman on 2021-09-02.
//

#include "GeoTools/GeoTile.h"
#include "GeoTools/GeoUtil.h"

#define _USE_MATH_DEFINES

#include <cmath>

using namespace AviationCalcUtil::GeoTools;
using namespace std;

GeoTile::GeoTile(const GeoPoint &point, double resolutionDegs) {
    // Get indexes for the specified resolution and floor it
    int maxLatIndex = static_cast<int>(floor(90 / resolutionDegs));
    int maxLonIndex = static_cast<int>(floor(180 / resolutionDegs));
    int latIndex = static_cast<int>(floor(point.getLat() / resolutionDegs));
    int lonIndex = static_cast<int>(floor(point.getLon() / resolutionDegs));

    // Check index
    if (latIndex == maxLatIndex) {
        latIndex--;
    }

    if (lonIndex == maxLonIndex) {
        lonIndex--;
    }

    // Set bounds
    botLat = latIndex * resolutionDegs;
    topLat = (latIndex + 1) * resolutionDegs;
    leftLon = lonIndex * resolutionDegs;
    rightLon = (lonIndex + 1) * resolutionDegs;
}

GeoTile::GeoTile(double botLat, double topLat, double leftLon, double rightLon) {
    if (rightLon < leftLon) {
        rightLon = leftLon;
    }

    if (topLat < botLat) {
        topLat = botLat;
    }

    this->botLat = fmin(fmax(botLat, -90), 90);
    this->topLat = fmin(fmax(topLat, -90), 90);
    this->leftLon = GeoUtil::normalizeLongitude(leftLon);
    this->rightLon = GeoUtil::normalizeLongitude(rightLon);
}

GeoTile::GeoTile(const GeoTile &tile) : GeoTile(tile.getBottomLat(), tile.getTopLat(), tile.getLeftLon(),
                                                tile.getRightLon()) {
}

bool GeoTile::isPointInTile(const GeoPoint &point) const {
    return point.getLat() >= botLat && point.getLat() <= topLat &&
           point.getLon() >= leftLon && point.getLon() <= rightLon;
}

bool GeoTile::operator==(const GeoPoint &point) const {
    return isPointInTile(point);
}

double GeoTile::getBottomLat() const {
    return botLat;
}

double GeoTile::getTopLat() const {
    return topLat;
}

double GeoTile::getLeftLon() const {
    return leftLon;
}

double GeoTile::getRightLon() const {
    return rightLon;
}

unique_ptr<GeoPoint> GeoTile::getCenterPoint() const {
    double centerLat = (botLat + topLat) / 2;
    double centerLon = (leftLon + rightLon) / 2;
    return make_unique<GeoPoint>(centerLat, centerLon);
}

File: src/geo/geo_tile.rs
use std::f64::consts::{FRAC_PI_2, PI};

use crate::units::{Length, Angle, Unit};

use super::{latitude::Latitude, longitude::Longitude, geo_point::GeoPoint};

pub trait GeoTile {
    fn bottom_lat(&self) -> Latitude;
    fn top_lat(&self) -> Latitude;
    fn left_lon(&self) -> Longitude;
    fn right_lon(&self) -> Longitude;

    fn center_point(&self) -> GeoPoint {
        let center_lat = Latitude::from_radians((self.bottom_lat().as_radians() + self.top_lat().as_radians()) / 2.0);
        let center_lon = Longitude::from_radians((self.left_lon().as_radians() + self.right_lon().as_radians()) / 2.0);

        return GeoPoint::new(center_lat, center_lon, Length::new(0.0));
    }

    fn is_point_in_tile(&self, point: &GeoPoint) -> bool {
        return point.lat >= self.bottom_lat() && point.lat <= self.top_lat() && point.lon >= self.left_lon() && point.lon <= self.right_lon();
    }
}

#[derive(Clone, Copy, Default, PartialEq, Debug)]
pub struct GeoTileBounds {
    bot_lat: Latitude,
    top_lat: Latitude,
    left_lon: Longitude,
    right_lon: Longitude
}

impl GeoTile for GeoTileBounds {
    fn bottom_lat(&self) -> Latitude {
        return self.bot_lat;
    }

    fn top_lat(&self) -> Latitude {
        return self.top_lat;
    }

    fn left_lon(&self) -> Longitude {
        return self.left_lon;
    }

    fn right_lon(&self) -> Longitude {
        return self.right_lon;
    }
}

impl GeoTileBounds {
    pub fn new(bot_lat: Latitude, mut top_lat: Latitude, left_lon: Longitude, mut right_lon: Longitude) -> GeoTileBounds {
        if (right_lon < left_lon){
            right_lon = left_lon;
        }

        if (top_lat < bot_lat){
            top_lat = bot_lat;
        }

        return GeoTileBounds { bot_lat, top_lat, left_lon, right_lon };
    }

    pub fn new_from_point(point: &GeoPoint, resolution: Angle) -> GeoTileBounds {
        let res = resolution.as_radians().abs();

        // Get indexes for the speciefied resolution and floor it
        let max_lat_index = (FRAC_PI_2 / res).floor() as i32;
        let max_lon_index = (PI / res).floor() as i32;
        let mut lat_index = (point.lat.as_radians() / res).floor() as i32;
        let mut lon_index = (point.lon.as_radians() / res).floor() as i32;

        // Check index
        if lat_index == max_lat_index {
            lat_index -= 1;
        }

        if lon_index == max_lon_index {
            lon_index -= 1;
        }

        return GeoTileBounds {
            bot_lat: Latitude::from_radians(lat_index as f64 * res),
            top_lat: Latitude::from_radians((lat_index + 1) as f64 * res),
            left_lon: Longitude::from_radians(lon_index as f64 * res),
            right_lon: Longitude::from_radians((lon_index + 1) as f64 * res)
        }
    }
}